applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py

```diff
@@ -0,0 +1,236 @@
+        # TODO : we need to modify UnionModelParts such that it does not give error
+        # remove this part after modifying UnionModelParts
+        if len(self.model_part_names)>1:
+            raise RuntimeError(f"A single model part is allowed")
+
+        self.model_part_name = self.model_part_names[0]
+        self.model_part = self.model.GetModelPart(self.model_part_name)
+
+        self.fixed_model_parts = self.parameters["settings"]["fixed_model_parts"].GetStringArray()
+        self.fixed_model_parts_thicknesses = self.parameters["settings"]["fixed_model_parts_thicknesses"].GetVector()
+        if len(self.fixed_model_parts) != len(self.fixed_model_parts_thicknesses):
+            raise RuntimeError(f"fixed_model_parts and fixed_model_parts_thicknesses should have the same size !")
+
+        helmholtz_settings = GetImplicitFilterParameters(self.parameters)
+
+        self.filter = HelmholtzAnalysis(self.model, helmholtz_settings)
+
+        self.controlled_physical_variable = Kratos.KratosGlobals.GetVariable("THICKNESS")
+
+    def Initialize(self) -> None:
+
+        # TODO. after solving issue with UnionModelParts such that it creates the union model part eventhough it exists
+        # we uncomment the following lines and work with the unioned nodel part
+        # model_parts_list = [self.model[model_part_name] for model_part_name in self.model_part_names]
+        # root_model_part = model_parts_list[0].GetRootModelPart()
+        # is_new_model_part, self.model_part = ModelPartUtilities.UnionModelParts(root_model_part, model_parts_list, False)
+
+        # if is_new_model_part:
+            # now create entity specific properties for the merged model part which is used for the control.
+        KratosOA.OptimizationUtils.CreateEntitySpecificPropertiesForContainer(self.model_part, self.model_part.Elements)
+
+        self.filter.Initialize()
+
+        # create the control field
+        self.control_field =  KratosOA.ContainerExpression.ElementPropertiesExpression(self.model_part)
+        self.control_field.SetData(self.parameters["settings"]["initial_thickness"].GetDouble())
+        self.control_field.Evaluate(self.controlled_physical_variable)
+
+    def Check(self):
+        return self.filter.Check()
+
+    def Finalize(self):
+        return self.filter.Finalize()
+
+    def GetPhysicalKratosVariables(self) -> list[SupportedSensitivityFieldVariableTypes]:
+        return [self.controlled_physical_variable]
+
+    def GetEmptyField(self) -> ContainerExpressionTypes:
+        field = KratosOA.ContainerExpression.ElementPropertiesExpression(self.model_part)
+        field.SetData(0.0)
+        return field
+
+    def GetControlField(self) -> ContainerExpressionTypes:
+        return self.control_field
+
+    def MapGradient(self, physical_gradient_variable_container_expression_map: dict[SupportedSensitivityFieldVariableTypes, ContainerExpressionTypes]) -> ContainerExpressionTypes:
```

Comment:
@sunethwarna I have noticed that here we use physical control to get the gradients,  
physical_gradient_variable_container_expression_map[self.controlled_physical_variable].Clone()

this is the same as the implementation in material control you implemented. The issue is from response_routine it is set like this. Do you have any idea to fix it ?!